### ISSUE TITLE
improve tutorials

### ISSUE THREAD

**Issue by hstojic** [secondmind-labs/trieste#472]
@vpicheny commented: 

I think that we've reached the point where we have too many notebooks (16 and growing) to have them in a single list! Also, titles are inconsistent, the lists are incomplete, etc.

Should we plan a lightweight refactor of the doc, including grouping the notebooks in subsections? any suggestion for subsections? In gpflow hey are grouped in Basic / Advanced / Tailor / Theory / Understanding, but I don't think it applies here...

**Comment by hstojic** [comment]
I was thinking the same thing few weeks ago and thought it would be good to have few more notebooks on basic stuff (how to setup a model, rules and acquisition functions, and some development related notebooks, e.g. how to add new models etc).

Here is my proposal for a new structure with sections, some of the notebooks here are missing - mostly in section 1 and 2

1. Quickstart (little explanation, how to jump in and get some results quickly - this could be re-used in the readme file at the root)

2. Introduction

    - Overview (what is BO - active learning as well?, basic formulation, some math and hints to more advanced things with links)
    - Data, search spaces and toy problems (illustrate datasets, observers and search spaces using our toy problems)
    - Modeling the objective function (why prob models, setting up models, build_model functions, priors, normalizing data, configs, DGP's and DNN's - some of these should probably go to an advanced modeling tutorial...)
    - Acquisition rules and functions (exploration-exploitation, mention different type of approaches like improvement, thompson sampling, differences between rules and builders in Trieste, optimization vs active learning, optimizer settings, discuss challenges and point to more advanced topics - batch, constraints, multi-objective)
    - Basic BO (current expected improvement essentially)
    - Basic active learning
    - Tensorboard
    - User in control of the BO loop (asktell inteface tutorial)

3. Standard algos  
    - EI
    - TS
    - LCB
    - MinValueEntropySearch
    - PredictiveVariance
    - ExpectedFeasibility and IVR

4. Advanced: Batch BO  
    - TS
    - BatchMonteCarloExpectedImprovement
    - MultipleOptimismNegativeLowerConfidenceBound
    - LocalPenalization
    - Fantasizer

5. Advanced: BO with constraints  
    - ExpectedConstrainedImprovement, ProbabilityOfFeasibility
    - etc

6. Advanced: Multi-objective BO  
    - ExpectedHypervolumeImprovement
    - BatchMonteCarloExpectedHypervolumeImprovement
    - etc

7. Miscellaneous  
    - normalising data
    - modeling configs
    - BO with deep ensembles
    - BO with deep GPs
    - Async BO
    - recovering from errors

8. Example applications  
    - Optimizing robotic controllers (OpenAI gym Lunar lander notebook

9. Developing with trieste  
    - How to add new models?
    - How to add new acquisition functions?

FAQ?  
- same as currently, various questions we think might arise, mostly mapping to the notebooks above

**Comment by vpicheny** [comment]
That looks great! would you see each point as an individual notebook? or is this the content that we should have, and we need to split this in fewer notebooks?

**Comment by vpicheny** [comment]
I think that section 2 addresses something important that is missing: we don't explain step by step all the important classes in `trieste`. This is because our introductory notebook uses a lot of the default configurations. For instance we don't specify the rule, nor the acquisition function. Perhaps we should make this purely a "quickstart" one, and create another one where we explicitely set step by step all of the elements?

**Comment by hstojic** [comment]
> That looks great! would you see each point as an individual notebook? or is this the content that we should have, and we need to split this in fewer notebooks?

Probably we can have fewer notebooks, different content can point to the same notebook (e.g. TS and batches with TS)

> I think that section 2 addresses something important that is missing: we don't explain step by step all the important classes in `trieste`. This is because our introductory notebook uses a lot of the default configurations. For instance we don't specify the rule, nor the acquisition function. Perhaps we should make this purely a "quickstart" one, and create another one where we explicitely set step by step all of the elements?

Yes, section 2 is the major one, right now we were trying to do some of that in few notebooks - explaining a bit mroe about mdoeling, priors etc. We should have dedicated intro sections and then illustrations of algos below, applications etc can be simplified and point to intro ones for details on basics.

**Comment by uri-granta** [comment]
Some initial, practical steps:

1. ¿Split notebooks across multiple pages? (currently just one page: Tutorials)
2. Split each page into multiple sections (currenty two: Examples and FAQ)
3. Give each notebook link an appropriate, short name (currently the full notebook title in Examples, a question in FAQ)
4. Decide which links should appear in the TOC on the left (currently all)

@hstojic How do you envisage your 9 sections fitting 1 and 2? Maybe have three pages: "Introduction" (sections 1-3), "Advanced" (sections 4-7) and "Development" (sections 8-9)? Do we still want an explicit FAQ section or do we expect people to be able to figure out the relevant section and link?

**Comment by heidariarash** [comment]
I would like to share my experience with using Trieste as a user as well.

Documentation is not enough at all. Notebooks are one thing, but the documentation should get improved as well. GP and BO are not easy-to-understand topics, so just providing the name and order of inputs in the documentation does not suffice. The same problem is also there in GPFlow documentation.

Important topics that need to be covered in the notebooks and also in the documentation in the first step:

1. How to define your own problem: All of the problems are solved with toy problems defined in the library, but not everyone is interested in solving DTLZ problems. I am using the library to solve my own problem, and there is no example/document regarding that.
2. How to define your own acquisition function: As a researcher, I am not always interested in using EI and EHVI. For somebody who is active in academia, it should be possible to define their own acquisition functions, for contributing to the science.
3. I guess this team is not working on GPFlow, but GPFlow models are an essential part of this library. I remember once I wanted to use Variational GP as a surrogate model, but apart from the name and the order of inputs in the documentation (sorry, not documentation, because there was no document for the models at that time. I found the related stuff in Github repo by browsing through different folders), and I finally gave up. It only said to use Z as inducing points. How should I define Z? just a NumPy array suffice?

and a lot more...

Consistency between different versions is another problem. I was a user since version 0.6.5, and everything changes for the next version, from simple things such as the order of inputs (the order of dataset and model changed last time when I was defining my own acquisition function, for no apparent reason), or the definition of a StackedModel for multi-objective optimization (every time, it is something different) to not so simple things (for example changing the directories in every version).

So far, I was using the Github repo (and not the documentation) to find what I need, which is not an easy task for somebody who is not a developer or familiar with Github, or I was using the help of my colleagues who are contributors to this repo.

Look at for example the documentation of Keras or SKLearn. Each documentation page has its own example, and that truly makes everything easier. Despite the fact that you can find whatever you want by just Googling something for Keras and Sci-Kit, their documentation is almost perfect. Neither BO and GP are as easy as the other machine learning methods nor does Trieste has a big community as these two. So, in my opinion, every detail should be included in the documentation.

I know documenting the code is not the easiest or the most exciting part of the job, but it is quite necessary, so I would be glad to help if needed.

**Comment by schmoelder** [comment]
Hey, 

I'm dipping my toes into trieste since it seems like a very powerful tool for BO (especially for constrained multi-objective BO). But I agree that the tutorials could be improved a bit. Most of my issues have been mentioned in this thread so I will not reiterate them and I really like the proposal for the new structure.

Just one thing I want to add, please make sure that users can directly use the tutorials without having to rely on any utility scripts. Even when running the very first example in your tutorials (https://secondmind-labs.github.io/trieste/notebooks/expected_improvement.html), it imports from `util.plotting_plotly` which was not found. I later found that this requires installing from source and running the notebooks in a specific directory. But first it took me some time to figure it out since I just wanted to start hacking and didn't really read the intro (I know, my bad). And secondly, it also adds a major overhead. As a new user, exploring different tools, I just want the tutorials to work out of the box so that I can better judge whether the tool fits my purpose or not.

**Comment by uri-granta** [comment]
@schmoelder Thanks for the feedback! The next release will have the plotting tools distributed as part of the package (https://github.com/secondmind-labs/trieste/pull/584), so the tutorials should work out of the box.

**Comment by vpicheny** [comment]
We're not documenting properly how to save / load models in trieste. We should do something about it! @uri-granta

**Comment by leeauk21** [comment]
does anyone has some code snippets for defining your own objective

**Comment by avivajpeyi** [comment]
Similar to @leeauk21 , im a bit confused how to define your own 'black box function'  that we can build a surrogate for using trieste. 

For example, 
```python

def black_box_function(x, y):
    return -x ** 2 - (y - 1) ** 2 + 1
```
